File: src/simulation/v0_9/python/parity.py
#!/usr/bin/env python3
"""Reference fixed-point inference used for Python/Bun parity fixtures."""

from __future__ import annotations

import argparse
import json
import math
from pathlib import Path
from typing import Any

INT16_MAX = 32767
INT32_MIN = -(2**31)
INT32_MAX = 2**31 - 1


def round_half_away(value: float) -> int:
    return math.floor(value + 0.5) if value >= 0 else math.ceil(value - 0.5)


def divide_half_away(value: int, divisor: int) -> int:
    magnitude = (abs(value) + divisor // 2) // divisor
    return magnitude if value >= 0 else -magnitude


def clamp(value: int, minimum: int, maximum: int) -> int:
    return max(minimum, min(maximum, value))


def score(artifact: dict[str, Any], raw_features: list[float]) -> int:
    architecture = artifact["architecture"]
    if len(raw_features) != architecture["inputSize"]:
        raise ValueError("parity feature width mismatch")
    normalization = artifact["normalization"]
    fixed = artifact["fixedPoint"]
    activations = []
    for index, raw in enumerate(raw_features):
        if not isinstance(raw, (int, float)) or not math.isfinite(raw):
            raise ValueError(f"feature {index} is not finite")
        normalized = (raw - normalization["offsets"][index]) * normalization["scales"][index]
        clipped = max(-fixed["inputClip"], min(fixed["inputClip"], normalized))
        activations.append(clamp(round_half_away(clipped * fixed["inputScale"]), -INT16_MAX, INT16_MAX))

    for layer in artifact["layers"]:
        if len(activations) != layer["inputSize"]:
            raise ValueError("dense layer shape mismatch")
        output = []
        divisor = 2 ** layer["scaleShift"]
        for row in range(layer["outputSize"]):
            accumulator = layer["biases"][row]
            offset = row * layer["inputSize"]
            for column, activation in enumerate(activations):
                accumulator += activation * layer["weights"][offset + column]
            value = divide_half_away(accumulator, divisor)
            if layer["activation"] == "relu":
                value = max(0, value)
                value = clamp(value, -INT16_MAX, INT16_MAX)
            else:
                value = clamp(value, INT32_MIN, INT32_MAX)
            output.append(value)
        activations = output
    return int(activations[0])


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--artifact", type=Path, required=True)
    parser.add_argument("--vectors", type=Path, required=True)
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    artifact = json.loads(args.artifact.read_text(encoding="utf-8"))
    if (
        artifact.get("schema") != "hoc.ai.v0_9_model.v1"
        or artifact.get("status") != "trained"
        or artifact.get("promoted") is not False
        or not artifact.get("modelSha256")
    ):
        raise ValueError("parity requires a sealed, unpromoted v0.9 research artifact")
    with args.vectors.open("r", encoding="utf-8") as stream:
        for line_number, line in enumerate(stream, start=1):
            if not line.strip():
                continue
            vector = json.loads(line)
            identifier = vector.get("id", f"line-{line_number}")
            value = score(artifact, vector["features"])
            expected = vector.get("expectedScore")
            if expected is not None and expected != value:
                raise ValueError(f"{identifier} expected {expected}, received {value}")
            print(json.dumps({"id": identifier, "score": value}, separators=(",", ":")))


if __name__ == "__main__":
    main()

File: src/simulation/v0_9/parity.ts
/*
 * -----------------------------------------------------------------------------
 * This file is part of the common code of the Heroes of Crypto.
 *
 * Heroes of Crypto and Heroes of Crypto AI are registered trademarks.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * -----------------------------------------------------------------------------
 */

import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
import { parseArgs } from "node:util";

import {
    scoreV09FixedPoint,
    serializeV09ModelHashPayload,
    validateV09ModelArtifact,
    type IV09ModelArtifact,
} from "../../ai/versions/v0_9_model";

export interface IV09ParityVector {
    id: string;
    features: number[];
    expectedScore?: number;
}

export interface IV09ParityResult {
    id: string;
    score: number;
}

const BASELINE_PARITY_VECTORS = 64;

const sha256 = (value: string): string => createHash("sha256").update(value).digest("hex");

export function verifyV09ResearchArtifact(artifact: IV09ModelArtifact): void {
    const errors = validateV09ModelArtifact(artifact);
    if (errors.length) throw new Error(`invalid v0.9 artifact: ${errors.join("; ")}`);
    if (artifact.status !== "trained" || artifact.promoted) {
        throw new Error("parity accepts an unpromoted trained research artifact");
    }
    if (!artifact.modelSha256 || sha256(serializeV09ModelHashPayload(artifact)) !== artifact.modelSha256) {
        throw new Error("v0.9 model hash mismatch");
    }
}

export function scoreV09ParityVectors(
    artifact: IV09ModelArtifact,
    vectors: readonly IV09ParityVector[],
): IV09ParityResult[] {
    verifyV09ResearchArtifact(artifact);
    return vectors.map((vector) => {
        if (!vector.id.trim()) throw new Error("parity vector id must be non-empty");
        const score = scoreV09FixedPoint(artifact, vector.features);
        if (vector.expectedScore !== undefined && vector.expectedScore !== score) {
            throw new Error(`parity vector ${vector.id} expected ${vector.expectedScore}, received ${score}`);
        }
        return { id: vector.id, score };
    });
}

/**
 * Corpus shared by the campaign parity gate. Besides broad deterministic values, it deliberately crosses input
 * half-rounding and clipping boundaries and drives every first-layer row toward its positive accumulator
 * envelope. Exact signed-int32 and post-accumulator half-rounding boundaries use sealed synthetic artifacts in
 * the focused test because an arbitrary learned model cannot be forced to land on those exact integers.
 */
export function buildV09ParityCorpus(artifact: IV09ModelArtifact): IV09ParityVector[] {
    verifyV09ResearchArtifact(artifact);
    const width = artifact.architecture.inputSize;
    const rawAtNormalized = (featureIndex: number, normalized: number): number => {
        const offset = artifact.normalization.offsets[featureIndex]!;
        const scale = artifact.normalization.scales[featureIndex]!;
        const raw = offset + normalized / scale;
        if (!Number.isFinite(raw)) {
            throw new Error(`v0.9 parity cannot represent boundary input for feature ${featureIndex}`);
        }
        return raw;
    };
    const atNormalized = (normalized: (featureIndex: number) => number): number[] =>
        Array.from({ length: width }, (_, featureIndex) => rawAtNormalized(featureIndex, normalized(featureIndex)));
    const vectors: IV09ParityVector[] = Array.from({ length: BASELINE_PARITY_VECTORS }, (_, vectorIndex) => ({
        id: `baseline-${vectorIndex}`,
        features: Array.from(
            { length: width },
            (_, featureIndex) => ((((vectorIndex + 1) * (featureIndex + 3)) % 37) - 18) / 7,
        ),
    }));
    const inputHalf = 0.5 / artifact.fixedPoint.inputScale;
    const beyondClip = artifact.fixedPoint.inputClip * 4;
    vectors.push(
        { id: "boundary-input-round-half-positive", features: atNormalized(() => inputHalf) },
        { id: "boundary-input-round-half-negative", features: atNormalized(() => -inputHalf) },
        { id: "boundary-input-clip-positive", features: atNormalized(() => beyondClip) },
        { id: "boundary-input-clip-negative", features: atNormalized(() => -beyondClip) },
        {
            id: "boundary-input-clip-alternating",
            features: atNormalized((featureIndex) => (featureIndex % 2 === 0 ? beyondClip : -beyondClip)),
        },
    );
    const firstLayer = artifact.layers[0]!;
    const rowCount = Math.min(firstLayer.outputSize, 8);
    for (let row = 0; row < rowCount; row += 1) {
        const rowOffset = row * firstLayer.inputSize;
        const biasSign = Math.sign(firstLayer.biases[row] ?? 0) || 1;
        vectors.push({
            id: `boundary-int32-accumulator-pressure-row-${row}`,
            features: atNormalized((featureIndex) => {
                const weightSign = Math.sign(firstLayer.weights[rowOffset + featureIndex] ?? 0);
                return beyondClip * (weightSign || biasSign);
            }),
        });
    }
    return vectors;
}

function parseVectors(contents: string): IV09ParityVector[] {
    return contents
        .split(/\r?\n/)
        .filter((line) => line.trim())
        .map((line, index) => {
            const value = JSON.parse(line) as Partial<IV09ParityVector>;
            if (!value || typeof value !== "object" || !Array.isArray(value.features)) {
                throw new Error(`invalid parity vector on line ${index + 1}`);
            }
            return {
                id: typeof value.id === "string" ? value.id : `line-${index + 1}`,
                features: value.features,
                ...(value.expectedScore === undefined ? {} : { expectedScore: value.expectedScore }),
            };
        });
}

function main(): void {
    const { values } = parseArgs({
        args: Bun.argv.slice(2),
        options: {
            artifact: { type: "string" },
            vectors: { type: "string" },
        },
        strict: true,
    });
    if (!values.artifact || !values.vectors) {
        throw new Error("usage: bun parity.ts --artifact <research.json> --vectors <vectors.jsonl>");
    }
    const artifact = JSON.parse(readFileSync(values.artifact, "utf8")) as IV09ModelArtifact;
    const results = scoreV09ParityVectors(artifact, parseVectors(readFileSync(values.vectors, "utf8")));
    for (const result of results) process.stdout.write(`${JSON.stringify(result)}\n`);
}

if (import.meta.main) main();
